Initial style guide (#1)
status-im/nim-style-guide: src/04_errors.md

# depending on Nim version and compiler flags) - name `XxxDefect`
type SomeDefect = object of Defect

# Use hierarchy for more specific errors
type MySpecificError = object of MyLibraryError

# Explicitly annotate functions with raises
func f() {.raises: [Defect, MySpecificError]} = discard

# Isolate code that may generate exceptions using expression-based try:
let x =
  try: ...
  except MyError as exc: ... # use the most specific error kind possible

# Be careful to catch excpetions inside loops, to avoid partial loop evaluations:
for x in y:
  try: ..
  except MyError: ..

# Provide contextual data when raising specific errors
raise (ref MyError)(msg: "description", data: value)
```

### Pros

* Used by `Nim` standard library
* Good for quick prototyping without error handling
* Good performance on happy path without `try`
  * Compatible with RVO

### Cons

* Poor readability - exceptions not part of API / signatures by default
    * Have to assume every line may fail
* Poor maintenance / refactoring support - compiler can't help detect affected code because they're not part of API
* Nim exception hierarchy unclear and changes between versions

>     * the distinction between `Exception`, `CatchableError` and `Defect` is inconsistently implemented

Reviewer: ```suggestion
    * The distinction between `Exception`, `CatchableError` and `Defect` is inconsistently implemented
```

        * [Exception hierarchy RFC not being implemented](https://github.com/nim-lang/Nim/issues/11776)
    * `Defect` is [not correctly tracked]((https://github.com/nim-lang/Nim/issues/12862))
    * Nim 1.4 further weakens compiler analysis around `Defect`(https://github.com/nim-lang/Nim/pull/13626)
* Without translation, exceptions leak information between abstraction layers

> * Writing exception-safe code in Nim unpractical due to missing critical features (compared to C++)

Reviewer 2: ```suggestion
* Writing exception-safe code in Nim impractical due to missing critical features present in C++
```

>     * no RAII - resources often leak in the presence of exceptions

Reviewer: ```suggestion
    * No RAII - resources often leak in the presence of exceptions
```

>     * destructors incomplete / unstable and thus not usable for safe EH

Reviewer: ```suggestion
    * Destructors incomplete / unstable and thus not usable for safe EH
```

>         * no constructors, thus no way to force particular object states at construction

Reviewer: ```suggestion
        * No constructors, thus no way to force particular object states at construction
```

    * `ref` types incompatible with destructors, even if they worked
* Poor performance of error path
    * Several heap allocations for each Exception (exception, stack trace, string)
    * Expensive stack trace
* Poor performance on happy path

>     * every `try` and `defer` has significant performance overhead due to `setjmp` exception handling implementation

Reviewer: ```suggestion
    * Every `try` and `defer` has significant performance overhead due to `setjmp` exception handling implementation
```

    * No way to transport accurate raises type information across Future/async/generic code boundaries - no `raisesof` equivalent of `typeof`

### Background

* [Stew EH helpers](https://github.com/status-im/nim-stew/pull/26) - Helpers that make working with checked exceptions easier
* [Nim Exception RFC](https://github.com/nim-lang/Nim/issues/8363) - seeks to differentiate between recoverable and unrecoverable errors
* [Zahary's handling proposal](https://gist.github.com/zah/d2d729b39d95a1dfedf8183ca35043b3) - seeks to handle any kind of error-generating API
* [C++ proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0709r0.pdf) - After 25 years of encouragement, half the polled C++ developers continue avoiding exceptions and Herb Sutter argues about the consequences of doing so
* [Google](https://google.github.io/styleguide/cppguide.html#Exceptions) and [llvm](https://llvm.org/docs/CodingStandards.html#id22) style guides on exceptions

## Status codes

Avoid status codes.

```nim

type StatusCode = enum
  Success
  Error1
  ...

func f(output: var Type): StatusCode
```

### Pros

* Interop with `C`

### Cons

* `output` undefined in case of error

> * verbose to use, must first declare mutable variable then call function and check result - mutable variable remains in scope even in "error" branch leading to bugs

Reviewer: ```suggestion
* Verbose to use, must first declare mutable variable then call function and check result - mutable variable remains in scope even in "error" branch leading to bugs
```
